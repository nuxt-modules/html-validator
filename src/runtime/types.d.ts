import { Result } from 'html-validate'
import { H3Event } from 'h3'
import { RenderResponse } from 'nitropack'

declare module 'nitropack' {
    interface NitroRuntimeHooks {
        'html-validator:check': (result: { valid: boolean, results: Result[] }, response: Partial<RenderResponse>, context: {
            event: H3Event;
        }) => void;
    }
}
