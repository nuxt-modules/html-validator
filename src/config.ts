import type { ConfigData } from 'html-validate'

export const defaultHtmlValidateConfig: ConfigData = {
  extends: [
    'html-validate:document',
    'html-validate:recommended',
    'html-validate:standard'
  ],
  rules: {
    //
    'svg-focusable': 'off',
    'no-unknown-elements': 'error',
    // Conflicts or not needed as we use prettier formatting
    'void-style': 'off',
    'no-trailing-whitespace': 'off',
    // Conflict with Nuxt defaults
    'require-sri': 'off',
    'attribute-boolean-style': 'off',
    'doctype-style': 'off',
    // Unreasonable rule
    'no-inline-style': 'off'
  }
}

export type LogLevel = 'verbose' | 'warning' | 'error'

export interface ModuleOptions {
  usePrettier?: boolean
  logLevel?: LogLevel
  failOnError?: boolean
  options?: ConfigData
}

export const DEFAULTS: Required<Omit<ModuleOptions, 'logLevel'>> & { logLevel?: LogLevel } = {
  usePrettier: false,
  failOnError: false,
  options: defaultHtmlValidateConfig
}
