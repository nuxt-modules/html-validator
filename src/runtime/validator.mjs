import chalk from 'chalk'
import { HtmlValidate, formatterFactory } from 'html-validate'

/** @param {import('html-validate').ConfigData} options */
export const getValidator = (options = {}) => {
  return new HtmlValidate(options)
}

/**
 * @param {import('html-validate').HtmlValidate} validator
 */
export const useChecker = (
  validator,
  usePrettier = false
) => {
  /** @type {string[]} */
  const invalidPages = []

  /**
   * @param {string} url
   * @param {string} html
   */
  const checkHTML = async (url, html) => {
    let couldFormat = false
    try {
      if (usePrettier) {
        const { format } = await import('prettier')
        html = format(html, { parser: 'html' })
        couldFormat = true
      }
    } catch (e) {
      console.error(e)
    }

    // Clean up Vue scoped style attributes
    html = typeof html === 'string' ? html.replace(/ ?data-v-[a-z0-9]+\b/g, '') : html
    const { valid, results } = validator.validateString(html)

    if (valid && !results.length) {
      return console.log(
        `No HTML validation errors found for ${chalk.bold(url)}`
      )
    }

    if (!valid) { invalidPages.push(url) }

    // TODO: investigate the many levels of default
    const formatter = couldFormat ? formatterFactory('codeframe') : await import('@html-validate/stylish').then(r => r.default?.default ?? r.default ?? r)

    const formattedResult = formatter?.(results)
    const reporter = valid ? console.warn : console.error

    reporter([
      `HTML validation errors found for ${chalk.bold(url)}`,
      formattedResult
    ].join('\n'))
  }

  return { checkHTML, invalidPages }
}
