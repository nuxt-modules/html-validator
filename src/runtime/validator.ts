import chalk from 'chalk'
import { ConfigData, HtmlValidate, formatterFactory } from 'html-validate'

export const getValidator = (options: ConfigData = {}) => {
  return new HtmlValidate(options)
}

export const useChecker = (
  validator: HtmlValidate,
  usePrettier = false
) => {
  const invalidPages: string[] = []

  const checkHTML = async (url: string, html: string) => {
    let couldFormat = false
    try {
      if (usePrettier) {
        const { format } = await import('prettier')
        html = format(html, { parser: 'html' })
        couldFormat = true
      }
    } catch (e) {
      console.error(e)
    }

    // Clean up Vue scoped style attributes
    html = typeof html === 'string' ? html.replace(/ ?data-v-[a-z0-9]+\b/g, '') : html
    const { valid, results } = validator.validateString(html)

    if (valid && !results.length) {
      return console.log(
        `No HTML validation errors found for ${chalk.bold(url)}`
      )
    }

    if (!valid) { invalidPages.push(url) }

    // TODO: investigate the many levels of default
    const formatter = couldFormat ? formatterFactory('codeframe') : await import('@html-validate/stylish').then(r => r.default?.default ?? r.default ?? r)

    const formattedResult = formatter?.(results)
    const reporter = valid ? console.warn : console.error

    reporter([
      `HTML validation errors found for ${chalk.bold(url)}`,
      formattedResult
    ].join('\n'))
  }

  return { checkHTML, invalidPages }
}
