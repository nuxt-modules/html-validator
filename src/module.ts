import { fileURLToPath, pathToFileURL } from 'node:url'
import chalk from 'chalk'
import { normalize } from 'pathe'
import { isWindows } from 'std-env'
import { genArrayFromRaw, genObjectFromRawEntries } from 'knitwork'

import { createResolver, defineNuxtModule, isNuxt2, logger, resolvePath } from '@nuxt/kit'
import { DEFAULTS } from './config'
import type { ModuleOptions } from './config'

export type { ModuleOptions }

export default defineNuxtModule<ModuleOptions>({
  meta: {
    name: '@nuxtjs/html-validator',
    configKey: 'htmlValidator',
    compatibility: {
      nuxt: '^2.0.0 || ^3.0.0-rc.7',
    },
  },
  defaults: nuxt => ({
    ...DEFAULTS,
    logLevel: nuxt.options.dev ? 'verbose' : 'warning',
  }),
  async setup(moduleOptions, nuxt) {
    logger.info(`Using ${chalk.bold('html-validate')} to validate server-rendered HTML`)

    const { usePrettier, failOnError, options, logLevel } = moduleOptions as Required<ModuleOptions>
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    if ((nuxt.options as any).htmlValidator?.options?.extends) {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      options.extends = (nuxt.options as any).htmlValidator.options.extends
    }

    const { resolve } = createResolver(import.meta.url)

    if (nuxt.options.dev) {
      nuxt.hook('nitro:config', (config) => {
        // Transpile the nitro plugin we're injecting
        config.externals = config.externals || {}
        config.externals.inline = config.externals.inline || []
        config.externals.inline.push('@nuxtjs/html-validator')

        // Add a nitro plugin that will run the validator for us on each request
        config.plugins = config.plugins || []
        config.plugins.push(normalize(fileURLToPath(new URL('./runtime/nitro', import.meta.url))))
        config.virtual = config.virtual || {}
        const serialisedOptions = genObjectFromRawEntries(Object.entries(moduleOptions).map(([key, value]) => {
          if (key !== 'ignore') return [key, JSON.stringify(value, null, 2)]
          const ignore = value as ModuleOptions['ignore'] || []
          return [key, genArrayFromRaw(ignore.map(v => typeof v === 'string' ? JSON.stringify(v) : v.toString()))]
        }))
        config.virtual['#html-validator-config'] = `export default ${serialisedOptions}`
      })

      nuxt.hook('prepare:types', ({ references }) => {
        const types = resolve('./runtime/types.d.ts')
        references.push({ path: types })
      })
    }

    if (!nuxt.options.dev || isNuxt2()) {
      const validatorPath = await resolvePath(fileURLToPath(new URL('./runtime/validator', import.meta.url)))
      const { useChecker, getValidator } = await import(isWindows ? pathToFileURL(validatorPath).href : validatorPath)
      const validator = getValidator(options)
      const { checkHTML, invalidPages } = useChecker(validator, usePrettier, logLevel)

      if (failOnError) {
        const errorIfNeeded = () => {
          if (invalidPages.length) {
            throw new Error('html-validator found errors')
          }
        }

        // @ts-expect-error TODO: use @nuxt/bridge-schema
        nuxt.hook('generate:done', errorIfNeeded)
        nuxt.hook('close', errorIfNeeded)
      }

      // Nuxt 3/Nuxt Bridge prerendering

      nuxt.hook('nitro:init', (nitro) => {
        nitro.hooks.hook('prerender:generate', (route) => {
          if (!route.contents || !route.fileName?.endsWith('.html')) {
            return
          }
          checkHTML(route.route, route.contents)
        })
      })

      // Nuxt 2

      if (isNuxt2()) {
        // @ts-expect-error TODO: use @nuxt/bridge-schema
        nuxt.hook('render:route', (url: string, result: { html: string }) => checkHTML(url, result.html))
        // @ts-expect-error TODO: use @nuxt/bridge-schema
        nuxt.hook('generate:page', ({ path, html }: { path: string, html: string }) => checkHTML(path, html))
      }
    }
  },
})
