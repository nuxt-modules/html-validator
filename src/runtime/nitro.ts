import type { NitroAppPlugin, RenderResponse } from 'nitropack'
import { getRequestPath } from 'h3'
import { useChecker, getValidator } from './validator'
// @ts-expect-error virtual module
import config from '#html-validator-config'

export default <NitroAppPlugin> function (nitro) {
  const validator = getValidator(config.options)
  const { checkHTML } = useChecker(validator, config.usePrettier, config.logLevel)

  nitro.hooks.hook('render:response', async (response: Partial<RenderResponse>, { event }) => {
    if (typeof response.body === 'string' && (response.headers?.['Content-Type'] || response.headers?.['content-type'])?.includes('html')) {
      // Block the response only if it's not hookable
      const promise = checkHTML(getRequestPath(event), response.body)
      if (config.hookable) {
        await nitro.hooks.callHook('html-validator:check', await promise, response, { event })
      }
    }
  })
}
