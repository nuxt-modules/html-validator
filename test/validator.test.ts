import { describe, it, expect } from 'vitest'
import { defaultHtmlValidateConfig } from '../src/config'
import { getValidator } from '../src/runtime/validator'

describe('useValidator', () => {
  it('generates a new validator for each set of options', () => {
    const option1 = { extends: [] }
    const validator1 = getValidator(option1)
    const validator2 = getValidator({ extends: ['html-validate:document'] })
    const validator3 = getValidator(option1)
    const validator4 = getValidator()
    const validator5 = getValidator()

    expect(validator1).not.toEqual(validator2)
    expect(validator1).toEqual(validator3)
    expect(validator4).toEqual(validator5)
  })

  it('returns a valid htmlValidate instance', () => {
    const validator = getValidator({ extends: ['html-validate:standard'] })

    const { valid, results } = validator.validateString('<!DOCTYPE html><title>x</title>')
    expect(valid).toBeTruthy()
    expect(results).toEqual([])

    const { valid: invalid, results: invalidResults } = validator.validateString('<!DOCTYPE html><title>x</title><body><a><a>Test</a></a></body>')
    expect(invalid).toBeFalsy()
    expect(invalidResults).toMatchSnapshot()
  })

  it('works with default config', () => {
    const validator = getValidator(defaultHtmlValidateConfig)
    const { valid, results } = validator.validateString('<!DOCTYPE html><title>x</title>')
    expect(valid).toBeTruthy()
    expect(results).toEqual([])
  })
})
