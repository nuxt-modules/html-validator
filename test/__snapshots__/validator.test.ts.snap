// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useValidator returns a valid htmlValidate instance 1`] = `
[
  {
    "errorCount": 1,
    "filePath": "inline",
    "messages": [
      {
        "column": 42,
        "context": {
          "ancestor": "<a>",
          "child": "<a>",
          "kind": "descendant",
        },
        "line": 1,
        "message": "<a> element is not permitted as a descendant of <a>",
        "offset": 41,
        "ruleId": "element-permitted-content",
        "ruleUrl": "https://html-validate.org/rules/element-permitted-content.html",
        "selector": "body > a > a",
        "severity": 2,
        "size": 1,
      },
    ],
    "source": "<!DOCTYPE html><title>x</title><body><a><a>Test</a></a></body>",
    "warningCount": 0,
  },
]
`;
